server/activate_hooks_test.go

```diff
@@ -77,8 +77,11 @@ func TestOnActivate(t *testing.T) {
 			},
 			ShouldError: true,
 		},
-		"minimum supported version fullfiled": {
+		"minimum supported version fulfilled": {
 			SetupAPI: func(api *plugintest.API) *plugintest.API {
+				api.On("KVSetWithOptions", mock.Anything, mock.Anything, mock.AnythingOfType("model.PluginKVSetOptions")).Return(true, nil).Maybe()
+				api.On("KVGet", mock.Anything).Return(nil, nil).Maybe()
```

Comment:
That is a to me the most biggest downside of the design: When using mocking for tests, the plugins has to know the implementation of the helper.

@levb I've seen some effort to split plugins into multiple packages. Do you think that approach can help getting rid of mocking?

Reply:
Yeah, I intentionally did the minimal amount of work necessary to get the tests to pass here. I think the right place to "fix" this in the plugin API rework, though I admit exactly how is still unknown. Just anything other than this ;P

Reply:
@hanzei We are moving towards wrapping all p.API dependencies in a consistent (for our purposes), plugin-specific interfaces. In the case of MS Calendar (WIP) you can see the implementation in https://github.com/levb/mattermost-plugin-mscalendar/blob/c8d6579d3324c5802c4f83c6533dd1c56c5ee614/server/mscalendar/mscalendar.go#L32-L38 and https://github.com/levb/mattermost-plugin-mscalendar/blob/lev-package-changes/server/utils/pluginapi/api.go.

The former defines the interface needed by MSCalendar; the latter provides a shareable "base" implementation, a go Type with a lot of methods. Then https://github.com/levb/mattermost-plugin-mscalendar/blob/c8d6579d3324c5802c4f83c6533dd1c56c5ee614/server/plugin/plugin.go#L232 ties them altogether.

Also, https://github.com/levb/mattermost-plugin-mscalendar/tree/lev-package-changes/server/utils/bot for all logs, messages, eventually webhooks etc. https://github.com/levb/mattermost-plugin-mscalendar/tree/lev-package-changes/server/utils is intended to be shareable, does not have any mscalendar dependencies.

This has been a result of slow evolution of our work, we are starting to converge across `workflow`, `mscalendar`, `autolink`, and `solar-lottery`. This layer may more useful for the more complex integration, and not intended as a replacement for the underlying "base" plugin API.

These specific interfaces are then easy to mock, as needed, and we use `gomock` for them.

cc @crspeller